newlib/libc/sys/vita/select.c

```diff
@@ -0,0 +1,128 @@
+to deal in the Software without restriction, including without limitation
+the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the
+Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+
+*/
+#include <sys/select.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <errno.h>
+
+#include <psp2/kernel/threadmgr.h>
+#include <psp2/net/net.h>
+#include <psp2/types.h>
+
+#include "vitadescriptor.h"
+
+#define SCE_ERRNO_MASK 0xFF
+
+#define MAX_EVENTS 255
+
+#define REG_POLL(_fd, event) \
+	do {														\
+		SceNetEpollEvent ev = {0};								\
+		ev.data.fd = (_fd);										\
+		ev.events = (event);									\
+		DescriptorTranslation *fdmap = __vita_fd_grab((_fd));	\
+		if (!fdmap) {											\
+			break;												\
+		}														\
+		sceNetEpollControl(eid, SCE_NET_EPOLL_CTL_ADD, fdmap->sce_uid, &ev);	\
+		__vita_fd_drop(fdmap);									\
+	} while(0)
+
+int select(int nfds, fd_set *readfds, fd_set *writefds, fd_set *exceptfds,
+           struct timeval *timeout)
+{
+	uint64_t wait = timeout->tv_sec * 1000000 + timeout->tv_usec;
+
+	if (nfds == 0) {
+		sceKernelDelayThread(wait);
+		return 0;
+	}
+
+	int i;
+	int res;
+	int eid = sceNetEpollCreate("", 0);
+	if (eid < 0) {
+		res = (int)eid;
+		goto exit;
+	}
+	for (i = 0; i < nfds; i++) {
+		if (readfds && FD_ISSET(i, readfds)) {
+			REG_POLL(i, SCE_NET_EPOLLIN);
+		}
+		if (writefds && FD_ISSET(i, writefds)) {
+			REG_POLL(i, SCE_NET_EPOLLOUT);
+		}
+		if (exceptfds && FD_ISSET(i, exceptfds)) {
+			REG_POLL(i, SCE_NET_EPOLLERR);
+		}
+	}
+	SceNetEpollEvent events[MAX_EVENTS] = {0};
+	int nev = sceNetEpollWait(eid, events, MAX_EVENTS, wait);
+
+	if (nev < 0) {
+		res = nev;
+		goto clean;
+	}
+
+	if (readfds) {
+		FD_ZERO(readfds);
+	}
+	if (writefds) {
+		FD_ZERO(writefds);
+	}
+	if (exceptfds) {
+		FD_ZERO(exceptfds);
+	}
+
+	res = 0;
+
+	for (i = 0; i < nev; i++) {
+		if (events[i].events) {
+			uint8_t inc = 0;
+			if (events[i].events & SCE_NET_EPOLLIN) {
+				inc = 1;
+				FD_SET(events[i].data.fd, readfds);
+			}
+			if (events[i].events & SCE_NET_EPOLLOUT) {
+				inc = 1;
+				FD_SET(events[i].data.fd, writefds);
+			}
+			if (events[i].events & SCE_NET_EPOLLERR) {
+				inc = 1;
+				FD_SET(events[i].data.fd, exceptfds);
+			}
+            res += inc;
```

Comment:
is this correct behavior? Or should it instead increment once per event?

http://pubs.opengroup.org/onlinepubs/9699919799/functions/select.html says

> Upon successful completion, the pselect() and select() functions shall return the total number of bits set in the bit masks. Otherwise, -1 shall be returned, and errno shall be set to indicate the error.

so I feel like the original behavior was correct here